cli/azd/pkg/templates/source_manager.go

```diff
@@ -0,0 +1,207 @@
+	"fmt"
+
+	"github.com/azure/azure-dev/cli/azd/pkg/config"
+	"github.com/azure/azure-dev/cli/azd/pkg/httputil"
+	"github.com/azure/azure-dev/cli/azd/resources"
+)
+
+const baseConfigKey string = "template.sources"
+
+var (
+	SourceDefault = &SourceConfig{
+		Key:  "default",
+		Name: "Default",
+		Type: SourceKindResource,
+	}
+
+	SourceAwesomeAzd = &SourceConfig{
+		Key:      "awesome-azd",
+		Name:     "Awesome AZD",
+		Type:     SourceKindAwesomeAzd,
+		Location: "https://aka.ms/awesome-azd/templates.json",
+	}
+
+	WellKnownSources = map[string]*SourceConfig{
+		SourceDefault.Key:    SourceDefault,
+		SourceAwesomeAzd.Key: SourceAwesomeAzd,
+	}
+
+	ErrSourceNotFound = errors.New("template source not found")
+	ErrSourceExists   = errors.New("template source already exists")
+)
+
+// SourceManager manages template sources used in azd template list and azd init experiences.
+type SourceManager interface {
+	// List returns a list of template sources.
+	List(ctx context.Context) ([]*SourceConfig, error)
+	// Get returns a template source by name.
+	Get(ctx context.Context, name string) (*SourceConfig, error)
+	// Add adds a new template source.
+	Add(ctx context.Context, key string, source *SourceConfig) error
+	// Remove removes a template source.
+	Remove(ctx context.Context, name string) error
+	// CreateSource creates a new template source from a source configuration
+	CreateSource(ctx context.Context, source *SourceConfig) (Source, error)
+}
+
+type sourceManager struct {
+	configManager config.UserConfigManager
+	httpClient    httputil.HttpClient
+}
+
+// NewSourceManager creates a new SourceManager.
+func NewSourceManager(configManager config.UserConfigManager, httpClient httputil.HttpClient) SourceManager {
+	return &sourceManager{
+		configManager: configManager,
+		httpClient:    httpClient,
+	}
+}
+
+// List returns a list of template sources.
+func (sm *sourceManager) List(ctx context.Context) ([]*SourceConfig, error) {
+	config, err := sm.configManager.Load()
+	if err != nil {
+		return nil, fmt.Errorf("unable to load user configuration: %w", err)
+	}
+
+	sourceConfigs := []*SourceConfig{}
+	rawSources, ok := config.Get(baseConfigKey)
+	if ok {
+		sourceMap := rawSources.(map[string]interface{})
+		for key, rawSource := range sourceMap {
+			var sourceConfig *SourceConfig
+
+			if wellKnownSource, ok := WellKnownSources[key]; ok {
+				sourceConfig = wellKnownSource
+			} else {
+				jsonBytes, err := json.Marshal(rawSource)
+				if err != nil {
+					return nil, fmt.Errorf("unable to parse source '%s': %w", key, err)
+				}
+
+				err = json.Unmarshal(jsonBytes, &sourceConfig)
+				if err != nil {
+					return nil, fmt.Errorf("unable to parse source '%s': %w", key, err)
+				}
+			}
+
+			sourceConfig.Key = key
+			sourceConfigs = append(sourceConfigs, sourceConfig)
+		}
+	}
+
+	// If not sources have been registered, add the default source.
+	if len(sourceConfigs) == 0 {
+		sourceConfigs = append(sourceConfigs, SourceDefault)
+	}
```

Comment:
Adding any template source removes the "Default" template source from showing up. Double checking -- this is indeed our intended behavior? I could see some users who would want to keep it, as some users would prefer it gone.

Reply:
We can iterate with @savannahostrowski to figure out which sources we want to enable to default and what would be intended here after users start adding their own sources.

Reply:
Can users remove sources that we've defined as default?

Reply:
> Can users remove sources that we've defined as default?

Not currently without adding some `disabled` feature to template sources.  So the idea would be to show our default sources until they add their own.  If they want the defaults back then they can manually add them.

Reply:
That feels a bit side effect-y to me. I think it'd probably feel a little nicer to retain the default sources unless the developer has consented/expressed the desire to remove them.

Reply:
Going to get this merged for now and we'll sort this out before next release.